src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java

```diff
@@ -245,23 +252,57 @@
         r.assertLogContains("1.0.0 > 1.2.0? false", b);
     }
 
-    @Test public void noReplayTrustedLibraries() throws Exception {
+    @Issue("JENKINS-41157")
+    @Test public void noReplayTrustedLibrariesWithoutRunScripts() throws Exception {
+        r.jenkins.setSecurityRealm(r.createDummySecurityRealm());
+        r.jenkins.setAuthorizationStrategy(new MockAuthorizationStrategy().
+                grantWithoutImplication(Jenkins.ADMINISTER).everywhere().to("bob").
+                grant(Jenkins.READ, Item.READ, View.READ).everywhere().to("bob"));
+
+        sampleRepo.init();
+        final String originalMessage = "must not be edited";
+        final String originalScript = "def call() {echo '" + originalMessage + "'}";
+        sampleRepo.write("vars/trusted.groovy", originalScript);
+        sampleRepo.git("add", "vars");
+        sampleRepo.git("commit", "--message=init");
+        GlobalLibraries.get().setLibraries(Collections.singletonList(new LibraryConfiguration("trusted", new SCMSourceRetriever(
+                new GitSCMSource(null, sampleRepo.toString(),
+                        "", "*", "", true)))));
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition("@Library('trusted@master') import trusted; trusted()", true));
+        final WorkflowRun b1 = r.buildAndAssertSuccess(p);
+        r.assertLogContains(originalMessage, b1);
+
+        ACL.impersonate(User.get("bob").impersonate(), new Runnable() {
+            @Override public void run() {
+                ReplayAction ra = b1.getAction(ReplayAction.class);
+                assertEquals(Collections.emptyMap(), ra.getOriginalLoadedScripts());
```

Comment:
This does not prove anything interesting, only that the untrusted user cannot use the web form to mount an attack.